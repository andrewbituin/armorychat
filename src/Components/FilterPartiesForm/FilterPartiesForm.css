.squads-filters__fieldset {
  border: none;
  padding: 0;
}

.squads-filters__fieldset legend {
  font-size: 14px;
  display: block;
  width: 100%;
  border-bottom: 1px solid #5b6370;
}

.squads-filters__fieldset input {
  margin: 15px 5px;
  width: calc(100% - 10px);
  border: 1px solid #6fa42c;
  background-color: transparent;
  border-radius: 5px;
  color: #fff;
  font-size: 16px !important;
  height: 30px;
  transition: background-color 0.6s, border 0.6s, box-shadow 0.6s;
  padding: 6px 8px 6px 32px;
}


.solo-input-transitions:focus, .solo-input-transitions:hover {
  box-shadow: 0 1px 8px 0 #00000080;
  /* border: 1px solid #565d69; */
  background-color: #212227;
}

.input-with-icon {
  position: relative;
  cursor: pointer;
}

.input-with-icon i {
  position: absolute;
  left: 14px;
  color: #b3b3b3;
  top: 1px;
}

.squads-filters-buttons-container {
  border-top: 1px solid #5b6370;
  display: flex;
  align-items: center;
  flex-direction: column;
}

.squads-filters-buttons-container button {
  margin-top: 15px;
  height: 25px;
  width: 90%;
}