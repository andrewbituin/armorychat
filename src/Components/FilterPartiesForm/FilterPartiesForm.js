import React, { useState, useContext } from 'react';
import GameContext from '../../Contexts/gameContext';
import Dropdown from '../Dropdown/Dropdown';
import './FilterPartiesForm.css'

export default function FilterPartyForm() {
  const gameContext = useContext(GameContext);
  const { requirements, roles, setFilters, resetFilters } = gameContext;
  const [searchTerm, setSearchTerm] = useState('');
  const [requirementFilter, setRequirementFilter] = useState(undefined);
  const [requirementFilter2, setRequirementFilter2] = useState(undefined);
  const [roleFilter, setRoleFilter] = useState(undefined);
  const [roleFilter2, setRoleFilter2] = useState(undefined);

  async function handleFilterSubmit(target) {
    const reqNodes = target.querySelectorAll('button[name=requirements]');
    let reqs = [];
    reqNodes.forEach(node => reqs.push(node.dataset.value));

    const roleNodes = target.querySelectorAll('button[name=roles]');
    let roles = [];
    roleNodes.forEach(node => roles.push(node.dataset.value));

    const searchTerm = target.querySelector('input').value;

    setFilters(searchTerm, reqs, roles);
  }
  
  function handleReset(e) {
    e.preventDefault();

    //check for differences, then update if changes found
    //controlled inputs
    searchTerm && setSearchTerm('');
    requirementFilter && setRequirementFilter(undefined);
    requirementFilter2 && setRequirementFilter2(undefined);
    roleFilter && setRoleFilter(undefined);
    roleFilter && setRoleFilter2(undefined);
    //context (live filters)
    resetFilters();
  }
  
  function handleFilterSubmitClick(e) {
    e.preventDefault();
    handleFilterSubmit(e.target);
  }

  function handleSearchChange(e) {
    e.preventDefault();
    setSearchTerm(e.target.value);
  }

  function generateRequirementFilterDropdown() {
    let temp = [];
    let options = {...requirements};

    temp.push(
      <Dropdown
        key={0}
        name="requirements"
        active={requirementFilter}
        onChange={e => setRequirementFilter(e.value)}
        onButtonClick={e => 
          {
            if (requirementFilter2) {
              setRequirementFilter(requirementFilter2);
              setRequirementFilter2(undefined)
            } else {
              setRequirementFilter(undefined); 
            }
          }
        }
        startValue={requirementFilter}
        placeholder='Select a requirement...'
        options={options}
      />
    );

    let tempOptions2 = {...options};
    delete tempOptions2[requirementFilter];

    temp.push(
      <Dropdown
        key={1}
        name="requirements"
        active={requirementFilter2}
        inactive={!requirementFilter}
        onChange={e => setRequirementFilter2(e.value)}
        onButtonClick={e => setRequirementFilter2(undefined)}
        startValue={requirementFilter2}
        placeholder='Select a requirement...'
        options={tempOptions2}
      />
    );

    return temp;
  }

  function generateRoleFilterDropdown() {
    let temp = [];
    let options = {...roles};

    temp.push(
      <Dropdown
        key={0}
        name="roles"
        active={roleFilter}
        gameId={gameContext.id}
        onChange={e => setRoleFilter(e.value)}
        onButtonClick={e => 
          {
            if (roleFilter2) {
              setRoleFilter(roleFilter2);
              setRoleFilter2(undefined)
            } else {
              setRoleFilter(undefined); 
            }
          }
        }
        startValue={roleFilter}
        placeholder='Select a role...'
        options={options}
      />
    );

    let tempOptions2 = {...options};
    delete tempOptions2[roleFilter];

    temp.push(
      <Dropdown
        key={1}
        name="roles"
        active={roleFilter2}
        inactive={!roleFilter}
        gameId={gameContext.id}
        onChange={e => setRoleFilter2(e.value)}
        onButtonClick={e => setRoleFilter2(undefined)}
        placeholder='Select a role...'
        startValue={roleFilter2}
        options={tempOptions2}
      />
    );

    return temp;
  }

  return (  
    <form className="squads-filters" onSubmit={handleFilterSubmitClick}>
      <fieldset className="squads-filters__fieldset">
        <legend>Search</legend>
        <label className="input-with-icon">
          <i className="fas fa-search relative-fieldset-icon"/>
          <input className="solo-input-transitions" type="text" onChange={handleSearchChange} placeholder="Search by name or description" value={searchTerm} maxLength="72"/>
        </label>
      </fieldset>
      <fieldset className="squads-filters__fieldset">
        <legend>Requirements</legend>
        {generateRequirementFilterDropdown()}
      </fieldset>
      <fieldset className="squads-filters__fieldset">
        <legend>Roles</legend>
        {generateRoleFilterDropdown()}
      </fieldset>
      <div className="squads-filters-buttons-container">
        <button className="green-button" type="submit">Filter Squads</button>
        <button className="grey-button" type="reset" onClick={handleReset}>Reset Filters</button>
      </div>
    </form>
  );
}