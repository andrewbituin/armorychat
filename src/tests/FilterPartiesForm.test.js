import React from 'react';
import FilterPartiesForm from '../Components/FilterPartiesForm/FilterPartiesForm'
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import renderer from "react-test-renderer";

describe('FilterPartiesForm component', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <FilterPartiesForm />
      </MemoryRouter>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  })

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <FilterPartiesForm />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})