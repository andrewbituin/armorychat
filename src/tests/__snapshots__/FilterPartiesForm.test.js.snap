// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterPartiesForm component renders the UI as expected 1`] = `
<form
  className="squads-filters"
  onSubmit={[Function]}
>
  <fieldset
    className="squads-filters__fieldset"
  >
    <legend>
      Search
    </legend>
    <label
      className="input-with-icon"
    >
      <i
        className="fas fa-search relative-fieldset-icon"
      />
      <input
        className="solo-input-transitions"
        maxLength="72"
        onChange={[Function]}
        placeholder="Search by name or description"
        type="text"
        value=""
      />
    </label>
  </fieldset>
  <fieldset
    className="squads-filters__fieldset"
  >
    <legend>
      Requirements
    </legend>
    <div
      className="dropdown"
    >
      <div
        className="css-1pcexqc-container"
        onKeyDown={[Function]}
      >
        <div
          className="css-0 r-dropdown__control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3 r-dropdown__value-container"
          >
            <div
              className="css-151xaom-placeholder r-dropdown__placeholder"
            >
              Select a requirement...
            </div>
            <input
              className="css-gj7qu5-dummyInput"
              disabled={false}
              id="react-select-4-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              readOnly={true}
              tabIndex="0"
              value=""
            />
          </div>
          <div
            className="css-1wy0on6 r-dropdown__indicators"
          >
            <span
              className="css-bgvzuu-indicatorSeparator r-dropdown__indicator-separator"
            />
            <div
              aria-hidden="true"
              className="css-16pqwjk-indicatorContainer r-dropdown__indicator r-dropdown__dropdown-indicator"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="inactive-dropdown"
    />
  </fieldset>
  <fieldset
    className="squads-filters__fieldset"
  >
    <legend>
      Roles
    </legend>
    <div
      className="dropdown"
    >
      <div
        className="css-1pcexqc-container"
        onKeyDown={[Function]}
      >
        <div
          className="css-0 r-dropdown__control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3 r-dropdown__value-container"
          >
            <div
              className="css-151xaom-placeholder r-dropdown__placeholder"
            >
              Select a role...
            </div>
            <input
              className="css-gj7qu5-dummyInput"
              disabled={false}
              id="react-select-5-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              readOnly={true}
              tabIndex="0"
              value=""
            />
          </div>
          <div
            className="css-1wy0on6 r-dropdown__indicators"
          >
            <span
              className="css-bgvzuu-indicatorSeparator r-dropdown__indicator-separator"
            />
            <div
              aria-hidden="true"
              className="css-16pqwjk-indicatorContainer r-dropdown__indicator r-dropdown__dropdown-indicator"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="inactive-dropdown"
    />
  </fieldset>
  <div
    className="squads-filters-buttons-container"
  >
    <button
      className="green-button"
      type="submit"
    >
      Filter Squads
    </button>
    <button
      className="grey-button"
      onClick={[Function]}
      type="reset"
    >
      Reset Filters
    </button>
  </div>
</form>
`;
